src/network/taraxa_capability.cpp

```diff
@@ -1416,58 +1416,129 @@ void TaraxaCapability::sendPbftBlocks(NodeID const &_id, size_t height_to_sync,
-    s.appendList(2);
-    s.appendRaw(level_0[i_0].rlp());
-    auto start_1 = edges_0_to_1[i_0];
-    auto end_1 = edges_0_to_1[i_0 + 1];
-    s.appendList(end_1 - start_1);
-    for (uint i_1 = start_1; i_1 < end_1; ++i_1) {
-      s.appendList(2);
-      s.appendRaw(level_1[i_1]);
-      auto start_2 = edges_1_to_2[i_1];
-      auto end_2 = edges_1_to_2[i_1 + 1];
-      s.appendList(end_2 - start_2);
-      for (uint i_2 = start_2; i_2 < end_2; ++i_2) {
-        s.appendRaw(level_2[i_2]);
+  auto transactions = db_query.execute();
+
+  // Creates final packet out of provided pbft blocks rlp representations
+  auto create_packet = [_id](std::vector<dev::bytes>&& pbft_blocks) -> RLPStream {
+    RLPStream packet_rlp;
+    packet_rlp.appendList(pbft_blocks.size());
+    for (const dev::bytes& block_rlp : pbft_blocks) {
+      packet_rlp.appendRaw(std::move(block_rlp));
+    }
+
+    return packet_rlp;
+  };
+
+  std::vector<dev::bytes> pbft_blocks_rlps;
+  uint64_t act_packet_size = 0;
+
+  // Iterate through pbft blocks
+  for (uint pbft_block_idx = 0; pbft_block_idx < pbft_blocks.size(); ++pbft_block_idx) {
+    RLPStream pbft_block_rlp;
+    auto start_1 = dag_blocks_indexes[pbft_block_idx];
+    auto end_1 = dag_blocks_indexes[pbft_block_idx + 1];
+
+    pbft_block_rlp.appendList(2); // item #1 - pbft block rlp, item #2 - list of dag blocks
+    pbft_block_rlp.appendRaw(pbft_blocks[pbft_block_idx].rlp());
+    pbft_block_rlp.appendList(end_1 - start_1);
+
+    // Iterate through dag blocks blocks / per pbft block
+    for (uint dag_block_idx = start_1; dag_block_idx < end_1; ++dag_block_idx) {
+      auto start_2 = transactions_indexes[dag_block_idx];
+      auto end_2 = transactions_indexes[dag_block_idx + 1];
+
+      pbft_block_rlp.appendList(2); // item #1 - dag block rlp, item #2 - list of dag block transactions
+      pbft_block_rlp.appendRaw(dag_blocks[dag_block_idx]);
+      pbft_block_rlp.appendList(end_2 - start_2);
+
+      // Iterate through txs / per dag block
+      for (uint trx_idx = start_2; trx_idx < end_2; ++trx_idx) {
+        pbft_block_rlp.appendRaw(transactions[trx_idx]);
+      }
+    }
+
+    // Check if PBFT blocks need to be split and sent in multiple packets so we dont exceed
+    // MAX_PACKET_SIZE (15 MB) limit
+    if (act_packet_size + pbft_block_rlp.out().size() > MAX_PACKET_SIZE) {
+      if (pbft_blocks_rlps.empty()) {
+        // If this log is present in log, it means there was created such big pbft block, that we are not able to send
+        // it through network
+        // This situation should never ever happen !!!
+        LOG(log_er_) << "Unable to send pbft block: " << pbft_blocks[pbft_block_idx].pbft_blk->getBlockHash().abridged()
+                     << " due to MAX_PACKET_SIZE(" << MAX_PACKET_SIZE << ") limit. "
+                     << "Block size: " << pbft_block_rlp.out().size() << " [B]";
+        // Skip this block as it would go through network anyway
```

Comment:
Think you mean to comment that it would NOT go through network anyway.   Perhaps we should kill the node if this occurs not just log an error?

Reply:
That block is already in db so killing/restarting node would solve nothing... But yeah it should not be even possible to create such pbft block so if it happens, we have a huge problem... Definitely needs some additional changes...